src/mon/commands/Command.h

```diff
@@ -0,0 +1,330 @@
+ *  Pending = the typename of the paxos service's pending map
+ *
+ * Depending on a paxos service's specificities, Stable and Pending may be
+ * the same.
+ *
+ * This class doesn't use these types for much more than defining the prepare
+ * and preprocess functions to be called by derived classes, but still
+ * requires them to be defined to properly abstract command definitions.
+ */
+
+typedef std::shared_ptr<Formatter> FormatterRef;
+
+template<typename Service, typename ...Args>
+struct Command : public Common<Service>
+{
+
+  CephContext *cct;
+
+  explicit Command(
+      Monitor *_mon,
+      Service *_service,
+      CephContext *_cct) :
+    Common<Service>(_mon, _service, _cct)
+  { 
+    cct = this->cct;
+  }
+
+  
+  virtual ~Command() { };
+
+  virtual bool preprocess(
+      MonOpRequestRef op,
+      const cmdmap_t &cmdmap,
+      Args ...args) {
+
+    stringstream ss;
+    bufferlist rdata;
+
+    string format;
+    cmd_getval(cct, cmdmap, "format", format, string("plain"));
+    FormatterRef f(Formatter::create(format));
+
+    string prefix;
+    cmd_getval(cct, cmdmap, "prefix", prefix);
+
+    return do_preprocess(op, prefix, cmdmap, ss, rdata, f, args...);
+  }
+
+  virtual bool prepare(
+      MonOpRequestRef op,
+      const cmdmap_t &cmdmap,
+      Args ...args) {
+
+    op->mark_osdmon_event(__func__);
+    stringstream ss;
+    bufferlist rdata;
+
+    string format;
+    cmd_getval(cct, cmdmap, "format", format, string("plain"));
+    FormatterRef f(Formatter::create(format));
+
+    string prefix;
+    cmd_getval(cct, cmdmap, "prefix", prefix);
+
+    return do_prepare(op, prefix, cmdmap,
+	ss, rdata, f, args...);
+  }
+
+  virtual bool do_preprocess(
+      MonOpRequestRef op,
+      const string &prefix,
+      const cmdmap_t &cmdmap,
+      stringstream &ss,
+      bufferlist rdata,
+      FormatterRef f,
+      Args ...args) = 0;
+
+  virtual bool do_prepare(
+      MonOpRequestRef op,
+      const string &prefix,
+      const cmdmap_t &cmdmap,
+      stringstream &ss,
+      bufferlist rdata,
+      FormatterRef f,
+      Args ...args) = 0;
+
+  virtual bool handles_command(const string &prefix) = 0;
+
+ private:
+  Command() = delete;
+};
+
+
+/* ReadCommand
+ *
+ * Defines a templated Read-only command. These will be called during
+ * preprocess. This class abstracts a derived class' command definition, by
+ * redefining the preprocess command from the 'Command' class, and then
+ * calling a function to be implemented by the class deriving from this one.
+ *
+ * Preparing is still left defined as the 'Command' class has it defined,
+ * which is basically returning 'false'.
+ */
+template<typename T, typename ...Args>
+struct ReadCommand : public Command<T, Args...>
+{
+
+  explicit ReadCommand(
+      Monitor *_mon,
+      T *_svc,
+      CephContext *_cct) :
+    Command<T, Args...>(_mon, _svc, _cct)
+  { };
+  virtual ~ReadCommand() { }
+
+  using Command<T, Args...>::prepare;
```

Comment:
Nice trick! :-)

There's a gadget that's been proposed in the standard called "override" that uses a technique like this with a variadic template-- you can unpack a parameter pack into a class-level "using" declaration, it's pretty slick.